### PR TITLE
15899 - affiliation invitations

### PR DESCRIPTION
**Pull request by ochiu**
*Issue #:*
https://github.com/bcgov/entity/issues/15899
*Description of changes:*
- Added new endpoints for affiliation invitations
- Basic email flow to send out affiliation invitation
- Pass code flow that creates and auto accepts an affiliation invitation

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the sbc-auth license (Apache 2.0).

### CODE REVIEW

File: auth-api/src/auth_api/models/affiliation_invitation.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+        """Calculate the expiry date based on the config value."""
+        if self.invitation_status_code == 'PENDING':
+            return self.sent_date + timedelta(minutes=int(get_named_config().AFFILIATION_TOKEN_EXPIRY_PERIOD_MINS))
+        return None
+
+    @hybrid_property
+    def status(self):
+        """Calculate the status based on the config value."""
+        current_time = datetime.now()
+        if self.invitation_status_code == 'PENDING':
```
Comment (seeker25) [inline auth-api/src/auth_api/models/affiliation_invitation.py:68]: Candidate for enum

File: auth-api/src/auth_api/models/affiliation_invitation.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+            expiry_time = self.sent_date + timedelta(
+                minutes=int(get_named_config().AFFILIATION_TOKEN_EXPIRY_PERIOD_MINS))
+            if current_time >= expiry_time:
+                return 'EXPIRED'
```
Comment (seeker25) [inline auth-api/src/auth_api/models/affiliation_invitation.py:72]: Candidate for enum

File: auth-api/src/auth_api/models/affiliation_invitation.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        return self.invitation_status_code
+
+    @classmethod
+    def create_from_dict(cls, invitation_info: dict, user_id, affiliation_id=None):
+        """Create a new Invitation from the provided dictionary."""
+        if invitation_info:
```
Comment (seeker25) [inline auth-api/src/auth_api/models/affiliation_invitation.py:78]:
```
if not invitation_info:
return None

```

File: auth-api/src/auth_api/models/affiliation_invitation.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+        return cls.query.filter_by(affiliation_id=affiliation_id).all()
+
+    @staticmethod
+    def find_invitations_by_org_entity_ids(from_org_id, to_org_id, entity_id):
+        """Find all affiliation invitation for org and entity ids."""
+        return db.session.query(AffiliationInvitation) \
+            .filter(AffiliationInvitation.from_org_id == from_org_id) \
+            .filter(AffiliationInvitation.to_org_id == to_org_id) \
+            .filter(AffiliationInvitation.entity_id == entity_id) \
+            .filter(or_(AffiliationInvitation.invitation_status_code == 'PENDING',
```
Comment (seeker25) [inline auth-api/src/auth_api/models/affiliation_invitation.py:144]: Use Enums

File: auth-api/src/auth_api/resources/affiliation_invitation.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    @_jwt.has_one_of_roles(
+        [Role.SYSTEM.value, Role.STAFF_VIEW_ACCOUNTS.value, Role.PUBLIC_USER.value])
+    def get():
+        """Get affiliation invitations."""
+        try:
+            org_id = request.args.get('orgId', None)
+            status = request.args.get('status', None)
+            data = AffiliationInvitationService.search_invitations_for_from_org(org_id, status)
+
+            if data is None:
```
Comment (seeker25) [inline auth-api/src/auth_api/resources/affiliation_invitation.py:51]: `data = data or []` maybe?

File: auth-api/src/auth_api/resources/affiliation_invitation.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    @TRACER.trace()
+    @cors.crossdomain(origin='*')
+    @_jwt.requires_auth
+    def put(affiliation_invitation_id, affiliation_invitation_token):
+        """Check whether the passed token is valid and add affiliation from the affiliation invitation."""
+        origin = request.environ.get('HTTP_ORIGIN', 'localhost')
+
+        try:
+            user = UserService.find_by_jwt_token()
+            if user is None:
```
Comment (seeker25) [inline auth-api/src/auth_api/resources/affiliation_invitation.py:154]: `if (user := UserService.find_by_jwt_token())`

File: auth-api/src/auth_api/utils/enums.py [hunk 7]
```diff
@@ -200,6 +200,13 @@ class InvitationType(Enum):
     STANDARD = 'STANDARD'  # Used to indicate the standard email invite with admin approval
 
 
+class AffiliationInvitationType(Enum):
+    """Affiliation Invitation type."""
+
+    EMAIL = 'EMAIL'  # Used to indicate an affiliation invitation initiated through email
```
Comment (seeker25) [inline auth-api/src/auth_api/utils/enums.py:206]: comment seems redundant? (just my opinion, could ignore this)

File: auth-api/tests/unit/models/test_affiliation_invitation.py [hunk 8]
```diff
@@ -0,0 +1,307 @@
+# Copyright © 2019 Province of British Columbia
```
Comment (seeker25) [inline auth-api/tests/unit/models/test_affiliation_invitation.py:1]: 2019 haha

File: auth-api/tests/unit/services/test_affiliation_invitation.py [hunk 9]
```diff
@@ -0,0 +1,379 @@
...
+from auth_api.services import Affiliation as AffiliationService
+from auth_api.services import AffiliationInvitation as AffiliationInvitationService
+from auth_api.services import Entity as EntityService
+from auth_api.services import Org as OrgService
+from auth_api.services import User
+from tests.utilities.factory_scenarios import TestEntityInfo, TestJwtClaims, TestOrgInfo, TestUserInfo
+from tests.utilities.factory_utils import factory_affiliation_invitation, factory_user_model, patch_token_info
+
+
+def create_test_entity():
```
Comment (seeker25) [inline auth-api/tests/unit/services/test_affiliation_invitation.py:38]: Good unit tests, thanks for these!

File: auth-api/src/auth_api/services/affiliation_invitation.py [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+        """Create a new affiliation invitation."""
+        context_path = CONFIG.AUTH_WEB_TOKEN_CONFIRM_PATH
+        from_org_id = affiliation_invitation_info['fromOrgId']
+        to_org_id = affiliation_invitation_info['toOrgId']
+        business_identifier = affiliation_invitation_info['businessIdentifier']
+
+        check_auth(org_id=from_org_id, one_of_roles=(ADMIN, COORDINATOR, STAFF))
+
+        # Validate from/to organizations exists
+        from_org: OrgModel = OrgModel.find_by_org_id(from_org_id)
```
Comment (seeker25) [inline auth-api/src/auth_api/services/affiliation_invitation.py:84]: `if not from_org := OrgModel.find_by_org_id(from_org_id): ` I think could work for most of these

File: auth-api/src/auth_api/services/affiliation_invitation.py [hunk 11]
```diff
@@ -0,0 +1,391 @@
...
+        return AffiliationInvitation(updated_invitation)
+
+    @staticmethod
+    def delete_affiliation_invitation(invitation_id):
+        """Delete the specified affiliation invitation."""
+        invitation: AffiliationInvitationModel = AffiliationInvitationModel.find_invitation_by_id(invitation_id)
+        if invitation is None:
+            raise BusinessException(Error.DATA_NOT_FOUND, None)
+
+        if invitation.status == 'ACCEPTED':
```
Comment (seeker25) [inline auth-api/src/auth_api/services/affiliation_invitation.py:184]: Enum

File: auth-api/src/auth_api/services/affiliation_invitation.py [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+
+    @staticmethod
+    @user_context
+    def search_invitations_for_from_org(org_id, status=None, **kwargs):
+        """Search affiliation invitations for from org."""
+        invitation_models = AffiliationInvitation.get_invitations_for_from_org(org_id, status, **kwargs)
+        invitations_result = {
+            'affiliationInvitations': []
+        }
+        for invitation in invitation_models:
```
Comment (seeker25) [inline auth-api/src/auth_api/services/affiliation_invitation.py:226]: Possible to use list comprehension for this?

Comment (seeker25) [reply auth-api/src/auth_api/services/affiliation_invitation.py:226]:
```
invitations_result['affiliationInvitations'] = [AffiliationInvitation(invitation).as_dict() for invitation in invitation_models]

```

File: auth-api/src/auth_api/services/affiliation_invitation.py [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+            })
+
+        return invitations_result
+
+    @staticmethod
+    @user_context
+    def get_invitations_for_to_org(org_id, status=None, **kwargs):
+        """Get affiliation invitations for to org."""
+        user_from_context: UserContext = kwargs['user_context']
+        org_model = OrgModel.find_by_org_id(org_id)
```
Comment (seeker25) [inline auth-api/src/auth_api/services/affiliation_invitation.py:238]: combine these two

File: auth-api/src/auth_api/services/affiliation_invitation.py [hunk 14]
```diff
@@ -0,0 +1,391 @@
...
+
+        return AffiliationInvitationModel.find_invitations_to_org(org_id=org_id, status=status)
+
+    @staticmethod
+    def find_affiliation_invitation_by_id(invitation_id):
+        """Find an existing affiliation invitation with the provided id."""
+        if invitation_id is None:
+            return None
+
+        invitation = AffiliationInvitationModel.find_invitation_by_id(invitation_id)
```
Comment (seeker25) [inline auth-api/src/auth_api/services/affiliation_invitation.py:266]: combine 266, 267

File: auth-api/src/auth_api/services/affiliation_invitation.py [hunk 15]
```diff
@@ -0,0 +1,391 @@
...
+
+        except Exception as e:  # noqa: E722
+            raise BusinessException(Error.EXPIRED_AFFILIATION_INVITATION, None) from e
+
+        affiliation_invitation: AffiliationInvitationModel = AffiliationInvitationModel.\
+            find_invitation_by_id(affiliation_invitation_id)
+
+        if affiliation_invitation is None:
+            raise BusinessException(Error.DATA_NOT_FOUND, None)
+        if affiliation_invitation.invitation_status_code == 'ACCEPTED':
```
Comment (seeker25) [inline auth-api/src/auth_api/services/affiliation_invitation.py:350]: enums

File: auth-api/src/auth_api/services/affiliation_invitation.py [hunk 16]
```diff
@@ -0,0 +1,391 @@
...
+                                      # pylint:disable=unused-argument
+                                      user: UserService, origin, **kwargs):
+        """Add an affiliation from the affilation invitation."""
+        current_app.logger.debug('>accept_affiliation_invitation')
+        affiliation_invitation: AffiliationInvitationModel = AffiliationInvitationModel.\
+            find_invitation_by_id(affiliation_invitation_id)
+
+        if affiliation_invitation is None:
+            raise BusinessException(Error.DATA_NOT_FOUND, None)
+        if affiliation_invitation.invitation_status_code == 'ACCEPTED':
```
Comment (seeker25) [inline auth-api/src/auth_api/services/affiliation_invitation.py:369]: enums

File: auth-api/src/auth_api/services/affiliation_invitation.py [hunk 17]
```diff
@@ -0,0 +1,391 @@
...
+            raise BusinessException(Error.ACTIONED_AFFILIATION_INVITATION, None)
+        if affiliation_invitation.invitation_status_code == 'EXPIRED':
+            raise BusinessException(Error.EXPIRED_AFFILIATION_INVITATION, None)
+
+        org_id = affiliation_invitation.to_org_id
+        entity_id = affiliation_invitation.entity_id
+
+        affiliation_model = AffiliationModel.find_affiliation_by_org_and_entity_ids(org_id, entity_id)
```
Comment (seeker25) [inline auth-api/src/auth_api/services/affiliation_invitation.py:377]: combine 377, 379

File: auth-api/src/auth_api/resources/affiliation_invitation.py (hunk 5 above)
Comment (Jxio) [reply auth-api/src/auth_api/resources/affiliation_invitation.py:51]: `search_invitations_for_from_org` function returns a dictionary with a key 'affiliationInvitations' that maps to a list.
maybe `data = data or {'affiliationInvitations': []}` ?

File: auth-api/src/auth_api/models/affiliation_invitation.py (hunk 1 above)
Comment (ochiu) [reply auth-api/src/auth_api/models/affiliation_invitation.py]: Updated

File: auth-api/src/auth_api/models/affiliation_invitation.py (hunk 2 above)
Comment (ochiu) [reply auth-api/src/auth_api/models/affiliation_invitation.py]: Updated

File: auth-api/src/auth_api/models/affiliation_invitation.py (hunk 3 above)
Comment (ochiu) [reply auth-api/src/auth_api/models/affiliation_invitation.py]: Added.

File: auth-api/src/auth_api/models/affiliation_invitation.py (hunk 4 above)
Comment (ochiu) [reply auth-api/src/auth_api/models/affiliation_invitation.py]: Updated

File: auth-api/src/auth_api/resources/affiliation_invitation.py (hunk 5 above)
Comment (ochiu) [reply auth-api/src/auth_api/resources/affiliation_invitation.py]: Good catches thanks! Updated as per Jia's comment

File: auth-api/src/auth_api/resources/affiliation_invitation.py (hunk 6 above)
Comment (ochiu) [reply auth-api/src/auth_api/resources/affiliation_invitation.py]: Updated

File: auth-api/src/auth_api/utils/enums.py (hunk 7 above)
Comment (ochiu) [reply auth-api/src/auth_api/utils/enums.py]: Removed comment

File: auth-api/tests/unit/models/test_affiliation_invitation.py (hunk 8 above)
Comment (ochiu) [reply auth-api/tests/unit/models/test_affiliation_invitation.py]: Updated :)

File: auth-api/src/auth_api/services/affiliation_invitation.py (hunk 10 above)
Comment (ochiu) [reply auth-api/src/auth_api/services/affiliation_invitation.py]: Cleaned this up and some of the subsequent checks to do this

File: auth-api/src/auth_api/services/affiliation_invitation.py (hunk 11 above)
Comment (ochiu) [reply auth-api/src/auth_api/services/affiliation_invitation.py]: Updated

File: auth-api/src/auth_api/services/affiliation_invitation.py (hunk 12 above)
Comment (ochiu) [reply auth-api/src/auth_api/services/affiliation_invitation.py]: Done, thanks for the suggestion

File: auth-api/src/auth_api/services/affiliation_invitation.py (hunk 13 above)
Comment (ochiu) [reply auth-api/src/auth_api/services/affiliation_invitation.py]: Updated

File: auth-api/src/auth_api/services/affiliation_invitation.py (hunk 14 above)
Comment (ochiu) [reply auth-api/src/auth_api/services/affiliation_invitation.py]: Updated

File: auth-api/src/auth_api/services/affiliation_invitation.py (hunk 15 above)
Comment (ochiu) [reply auth-api/src/auth_api/services/affiliation_invitation.py]: Updated

File: auth-api/src/auth_api/services/affiliation_invitation.py (hunk 16 above)
Comment (ochiu) [reply auth-api/src/auth_api/services/affiliation_invitation.py]: Updated

File: auth-api/src/auth_api/services/affiliation_invitation.py (hunk 17 above)
Comment (ochiu) [reply auth-api/src/auth_api/services/affiliation_invitation.py]: Updated

File: auth-api/src/auth_api/services/affiliation_invitation.py (hunk 12 above)
Comment (seeker25) [reply auth-api/src/auth_api/services/affiliation_invitation.py]: I used chatgpt for this suggestion - could always just throw some code in there - ask it to optimize

**Review by seeker25 — APPROVED** [review]
High level looks good, I'll spend some time after this is merged today - hitting it with postman with different security permissions

**Pull request merged**